File: apps/ts/src/routes/store.ts
/**
 * Store API routes for the Fractal Agents Runtime — TypeScript/Bun.
 *
 * Implements LangGraph-compatible cross-thread key-value storage endpoints
 * matching the Python runtime:
 *
 *   PUT    /store/items         — Store/update (upsert) an item
 *   GET    /store/items         — Retrieve an item by namespace + key
 *   DELETE /store/items         — Delete an item by namespace + key
 *   POST   /store/items/search  — Search items in a namespace
 *   GET    /store/namespaces    — List namespaces for the authenticated user
 *
 * Response conventions (matching Python exactly):
 *   - PUT returns 200 with the StoreItem object.
 *   - GET returns 200 with the StoreItem, or 404 if not found.
 *   - DELETE returns 200 with `{}`, or 404 if not found.
 *   - Search returns 200 with a JSON array of StoreItem objects.
 *   - Namespaces returns 200 with a JSON array of namespace strings.
 *   - Errors use `{"detail": "..."}` shape (ErrorResponse).
 *
 * All operations are scoped by the authenticated user's identity. When
 * auth is disabled (Supabase not configured), `ownerId` defaults to
 * `"anonymous"` so the store still works in development.
 *
 * Reference: apps/python/src/server/routes/store.py
 */

import type { Router } from "../router";
import type { StorePutRequest, StoreSearchRequest } from "../models/store";
import {
  jsonResponse,
  validationError,
  requireBody,
  notFound,
} from "./helpers";
import { getStorage } from "../storage/index";
import { getUserIdentity } from "../middleware/context";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/**
 * Default owner ID when authentication is disabled.
 *
 * Ensures the store is usable in development without Supabase. All
 * unauthenticated requests share the same "anonymous" namespace.
 */
const ANONYMOUS_OWNER_ID = "anonymous";

// ---------------------------------------------------------------------------
// Route registration
// ---------------------------------------------------------------------------

/**
 * Register all store routes on the given router.
 *
 * @param router - The application Router instance.
 */
export function registerStoreRoutes(router: Router): void {
  // IMPORTANT: Register /store/items/search BEFORE /store/items so that
  // "search" is matched as a literal path segment, not treated as part
  // of the /store/items pattern.
  router.post("/store/items/search", handleSearchStoreItems);

  router.put("/store/items", handlePutStoreItem);
  router.get("/store/items", handleGetStoreItem);
  router.delete("/store/items", handleDeleteStoreItem);

  router.get("/store/namespaces", handleListNamespaces);
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Get the effective owner ID for the current request.
 *
 * Returns the authenticated user's identity, or "anonymous" when auth
 * is disabled.
 */
function getEffectiveOwnerId(): string {
  return getUserIdentity() ?? ANONYMOUS_OWNER_ID;
}

// ---------------------------------------------------------------------------
// PUT /store/items — Store or update an item
// ---------------------------------------------------------------------------

/**
 * Handle `PUT /store/items`.
 *
 * Upserts a store item by `(namespace, key)`. If the item already exists
 * for the authenticated user, its value and updated_at are overwritten.
 *
 * Request body: `{ namespace: string, key: string, value: object, metadata?: object }`
 * Response: StoreItem (200) or validation error (422)
 */
async function handlePutStoreItem(
  request: Request,
  _params: Record<string, string>,
  _query: URLSearchParams,
): Promise<Response> {
  const [body, bodyError] = await requireBody<StorePutRequest>(request);
  if (bodyError) return bodyError;

  // Validate required fields
  const { namespace, key, value, metadata } = body;

  if (!namespace || typeof namespace !== "string") {
    return validationError("namespace is required");
  }
  if (!key || typeof key !== "string") {
    return validationError("key is required");
  }
  if (value === undefined || value === null) {
    return validationError("value is required");
  }

  const ownerId = getEffectiveOwnerId();
  const storage = getStorage();

  const item = await storage.store.put(
    namespace,
    key,
    typeof value === "object" ? (value as Record<string, unknown>) : { _value: value },
    ownerId,
    metadata,
  );

  return jsonResponse(item);
}

// ---------------------------------------------------------------------------
// GET /store/items — Get an item by namespace + key
// ---------------------------------------------------------------------------

/**
 * Handle `GET /store/items`.
 *
 * Retrieves a store item by namespace and key, provided as query parameters.
 *
 * Query params: `?namespace=<string>&key=<string>`
 * Response: StoreItem (200) or 404
 */
async function handleGetStoreItem(
  _request: Request,
  _params: Record<string, string>,
  query: URLSearchParams,
): Promise<Response> {
  const namespace = query.get("namespace");
  const key = query.get("key");

  if (!namespace) {
    return validationError("namespace query parameter is required");
  }
  if (!key) {
    return validationError("key query parameter is required");
  }

  const ownerId = getEffectiveOwnerId();
  const storage = getStorage();

  const item = await storage.store.get(namespace, key, ownerId);

  if (item === null) {
    return notFound(`Item not found: ${namespace}/${key}`);
  }

  return jsonResponse(item);
}

// ---------------------------------------------------------------------------
// DELETE /store/items — Delete an item by namespace + key
// ---------------------------------------------------------------------------

/**
 * Handle `DELETE /store/items`.
 *
 * Deletes a store item by namespace and key, provided as query parameters.
 *
 * Query params: `?namespace=<string>&key=<string>`
 * Response: `{}` (200) or 404
 */
async function handleDeleteStoreItem(
  _request: Request,
  _params: Record<string, string>,
  query: URLSearchParams,
): Promise<Response> {
  const namespace = query.get("namespace");
  const key = query.get("key");

  if (!namespace) {
    return validationError("namespace query parameter is required");
  }
  if (!key) {
    return validationError("key query parameter is required");
  }

  const ownerId = getEffectiveOwnerId();
  const storage = getStorage();

  const deleted = await storage.store.delete(namespace, key, ownerId);

  if (!deleted) {
    return notFound(`Item not found: ${namespace}/${key}`);
  }

  return jsonResponse({});
}

// ---------------------------------------------------------------------------
// POST /store/items/search — Search items in a namespace
// ---------------------------------------------------------------------------

/**
 * Handle `POST /store/items/search`.
 *
 * Searches items within a namespace, optionally filtering by key prefix.
 * Results are sorted by key and paginated.
 *
 * Request body: `{ namespace: string, prefix?: string, limit?: number, offset?: number }`
 * Response: StoreItem[] (200)
 */
async function handleSearchStoreItems(
  request: Request,
  _params: Record<string, string>,
  _query: URLSearchParams,
): Promise<Response> {
  const [body, bodyError] = await requireBody<StoreSearchRequest>(request);
  if (bodyError) return bodyError;

  const { namespace, prefix } = body;

  if (!namespace || typeof namespace !== "string") {
    return validationError("namespace is required");
  }

  // Validate and clamp pagination params
  let limit = 10;
  let offset = 0;

  if (body.limit !== undefined) {
    const parsedLimit = Number(body.limit);
    if (Number.isNaN(parsedLimit)) {
      return validationError("limit must be a number");
    }
    limit = Math.max(1, Math.min(parsedLimit, 100));
  }

  if (body.offset !== undefined) {
    const parsedOffset = Number(body.offset);
    if (Number.isNaN(parsedOffset)) {
      return validationError("offset must be a number");
    }
    offset = Math.max(0, parsedOffset);
  }

  const ownerId = getEffectiveOwnerId();
  const storage = getStorage();

  const items = await storage.store.search(
    namespace,
    ownerId,
    prefix,
    limit,
    offset,
  );

  return jsonResponse(items);
}

// ---------------------------------------------------------------------------
// GET /store/namespaces — List namespaces for the authenticated user
// ---------------------------------------------------------------------------

/**
 * Handle `GET /store/namespaces`.
 *
 * Returns all namespace strings that contain at least one item for the
 * authenticated user.
 *
 * Response: string[] (200)
 */
async function handleListNamespaces(
  _request: Request,
  _params: Record<string, string>,
  _query: URLSearchParams,
): Promise<Response> {
  const ownerId = getEffectiveOwnerId();
  const storage = getStorage();

  const namespaces = await storage.store.listNamespaces(ownerId);

  return jsonResponse(namespaces);
}

File: apps/python/src/server/routes/store.py
"""Store API routes for Robyn server.

Implements LangGraph-compatible key-value storage endpoints:
- PUT /store/items — Store/update items
- GET /store/items — Retrieve items by namespace/key
- DELETE /store/items — Delete items
- POST /store/items/search — Search items by prefix
- GET /store/namespaces — List namespaces
"""

from __future__ import annotations

import json
import logging
from urllib.parse import unquote

from robyn import Request, Response, Robyn

from server.auth import AuthenticationError, require_user
from server.routes.helpers import error_response, json_response, parse_json_body
from server.storage import get_storage

logger = logging.getLogger(__name__)


def _normalise_namespace(namespace: str | list[str] | None) -> str | None:
    """Normalise a namespace value to a dot-joined string.

    The LangGraph SDK convention sends namespaces as ``list[str]`` (tuple of
    path segments) in JSON bodies (PUT, search).  Query parameters (GET,
    DELETE) arrive as plain strings — or as JSON-encoded arrays when sent
    by k6 / SDK clients.

    This helper accepts both forms and returns a canonical dot-joined
    string, matching the TS runtime's ``normaliseNamespace()`` from
    ``routes/store.ts``.

    Args:
        namespace: A string, list of strings, or ``None`` from the request.

    Returns:
        The canonical dot-joined string, or ``None`` if the value is
        empty / invalid.
    """
    if isinstance(namespace, list):
        segments = [s for s in namespace if isinstance(s, str) and len(s) > 0]
        return ".".join(segments) if segments else None
    if isinstance(namespace, str) and len(namespace) > 0:
        # Robyn does NOT URL-decode query parameter values, so a k6/SDK
        # request like ``namespace=%5B%22a%22%2C%22b%22%5D`` arrives as
        # the raw percent-encoded string.  URL-decode first, then try
        # JSON-parse in case it's a JSON-encoded array (e.g. '["a","b"]').
        url_decoded = unquote(namespace)
        try:
            decoded = json.loads(url_decoded)
            if isinstance(decoded, list):
                return _normalise_namespace(decoded)
        except (json.JSONDecodeError, TypeError):
            pass
        return url_decoded
    return None


def register_store_routes(app: Robyn) -> None:
    """Register store routes with the Robyn app.

    Args:
        app: Robyn application instance
    """

    # ========================================================================
    # Store Items - CRUD Operations
    # ========================================================================

    @app.put("/store/items")
    async def put_store_item(request: Request) -> Response:
        """Store or update an item.

        Request body:
        {
            "namespace": "string",
            "key": "string",
            "value": any,
            "metadata": {"key": "value"}  // optional
        }

        Response: StoreItem (200) or error (4xx)
        """
        try:
            user = require_user()
        except AuthenticationError as e:
            return error_response(e.message, 401)

        try:
            body = parse_json_body(request)
        except json.JSONDecodeError:
            return error_response("Invalid JSON in request body", 422)

        # Validate required fields
        namespace = _normalise_namespace(body.get("namespace"))
        key = body.get("key")
        value = body.get("value")

        if not namespace:
            return error_response("namespace is required", 422)
        if not key:
            return error_response("key is required", 422)
        if value is None:
            return error_response("value is required", 422)

        metadata = body.get("metadata")

        storage = get_storage()
        item = await storage.store.put(
            namespace=namespace,
            key=key,
            value=value,
            owner_id=user.identity,
            metadata=metadata,
        )

        return json_response(item.to_dict())

    @app.get("/store/items")
    async def get_store_item(request: Request) -> Response:
        """Get an item by namespace and key.

        Query params:
        - namespace: Namespace (required)
        - key: Key within namespace (required)

        Response: StoreItem (200) or error (404)
        """
        try:
            user = require_user()
        except AuthenticationError as e:
            return error_response(e.message, 401)

        # Parse query params — namespace may be a plain string or a
        # JSON-encoded array (e.g. '["benchmark","ts"]' from k6/SDK).
        raw_namespace = None
        key = None

        if request.query_params:
            raw_namespace = request.query_params.get("namespace", None)
            key = request.query_params.get("key", None)

        namespace = _normalise_namespace(raw_namespace)
        if not namespace:
            return error_response("namespace query parameter is required", 422)
        if not key:
            return error_response("key query parameter is required", 422)

        storage = get_storage()
        item = await storage.store.get(
            namespace=namespace,
            key=key,
            owner_id=user.identity,
        )

        if item is None:
            return error_response(f"Item not found: {namespace}/{key}", 404)

        return json_response(item.to_dict())

    @app.delete("/store/items")
    async def delete_store_item(request: Request) -> Response:
        """Delete an item by namespace and key.

        Query params:
        - namespace: Namespace (required)
        - key: Key within namespace (required)

        Response: empty object (200) or error (404)
        """
        try:
            user = require_user()
        except AuthenticationError as e:
            return error_response(e.message, 401)

        # Parse query params — same normalisation as GET.
        raw_namespace = None
        key = None

        if request.query_params:
            raw_namespace = request.query_params.get("namespace", None)
            key = request.query_params.get("key", None)

        namespace = _normalise_namespace(raw_namespace)
        if not namespace:
            return error_response("namespace query parameter is required", 422)
        if not key:
            return error_response("key query parameter is required", 422)

        storage = get_storage()
        deleted = await storage.store.delete(
            namespace=namespace,
            key=key,
            owner_id=user.identity,
        )

        if not deleted:
            return error_response(f"Item not found: {namespace}/{key}", 404)

        return json_response({})

    # ========================================================================
    # Store Items - Search
    # ========================================================================

    @app.post("/store/items/search")
    async def search_store_items(request: Request) -> Response:
        """Search items in a namespace.

        Request body:
        {
            "namespace": "string",
            "prefix": "string",  // optional key prefix filter
            "limit": 10,         // optional, default 10
            "offset": 0          // optional, default 0
        }

        Response: list[StoreItem] (200)
        """
        try:
            user = require_user()
        except AuthenticationError as e:
            return error_response(e.message, 401)

        try:
            body = parse_json_body(request)
        except json.JSONDecodeError:
            return error_response("Invalid JSON in request body", 422)

        namespace = _normalise_namespace(body.get("namespace"))
        if not namespace:
            return error_response("namespace is required", 422)

        prefix = body.get("prefix")
        limit = body.get("limit", 10)
        offset = body.get("offset", 0)

        # Validate pagination
        try:
            limit = max(1, min(int(limit), 100))
            offset = max(0, int(offset))
        except (TypeError, ValueError):
            return error_response("limit and offset must be integers", 422)

        storage = get_storage()
        items = await storage.store.search(
            namespace=namespace,
            owner_id=user.identity,
            prefix=prefix,
            limit=limit,
            offset=offset,
        )

        return json_response([item.to_dict() for item in items])

    # ========================================================================
    # Store Namespaces
    # ========================================================================

    @app.get("/store/namespaces")
    async def list_namespaces(request: Request) -> Response:
        """List all namespaces for the current user.

        Response: list[string] (200)
        """
        try:
            user = require_user()
        except AuthenticationError as e:
            return error_response(e.message, 401)

        storage = get_storage()
        namespaces = await storage.store.list_namespaces(user.identity)

        return json_response(namespaces)
